platform/sync-installer/sync-installer-impl/src/main/java/org/codice/ddf/sync/installer/impl/SynchronisedInstallerImpl.java

```diff
@@ -0,0 +1,566 @@
+      throw new SynchronisedInstallerException(
+          "Bundles failed to start ["
+              + failedBundles.stream().map(Bundle::getSymbolicName).collect(Collectors.joining(","))
+              + "].");
+    }
+
+    if (!unavailableBundles.isEmpty()) {
+      LOGGER.info(
+          "Waiting on the following bundles to reach Active state [{}]",
+          unavailableBundles
+              .stream()
+              .map(Bundle::getSymbolicName)
+              .collect(Collectors.joining(",")));
+    }
+
+    return unavailableBundles;
+  }
+
+  @VisibleForTesting
+  protected ServiceTracker getServiceTracker(BundleContext bundleContext, Filter filter) {
+    return new ServiceTracker<>(bundleContext, filter, null);
+  }
+
+  private void printBundleInformation(
+      BiConsumer<String, Object[]> logConsumer, Collection<Bundle> bundles) {
+    for (Bundle bundle : bundles) {
+      StringBuilder headerString = new StringBuilder("[ ");
+      Enumeration<String> keys = bundle.getHeaders().keys();
+
+      while (keys.hasMoreElements()) {
+        String key = keys.nextElement();
+        headerString.append(key).append("=").append(bundle.getHeaders().get(key)).append(", ");
+      }
+
+      headerString.append(" ]");
+      logConsumer.accept(
+          "\n\tBundle: {}_v{} | {}\n\tHeaders: {}",
+          new Object[] {
+            bundle.getSymbolicName(),
+            bundle.getVersion(),
+            BUNDLE_STATES.getOrDefault(bundle.getState(), "UNKNOWN"),
+            headerString
+          });
+    }
+  }
+
+  private Stream<Feature> getFeaturesFromNames(String feature, String... additionalFeatures)
+      throws SynchronisedInstallerException {
+    Stream.Builder<Feature> transformedFeatures = Stream.builder();
+
+    for (String f : toList(feature, additionalFeatures)) {
+      try {
+        transformedFeatures.add(featuresService.getFeature(f));
+      } catch (Exception e) {
+        throw new SynchronisedInstallerException("Failed to retrieve feature [" + f + "]", e);
+      }
+    }
+
+    return transformedFeatures.build();
+  }
+
+  private Filter getServiceIdFilter(String servicePid) throws SynchronisedInstallerException {
+    try {
+      return bundleContext.createFilter(String.format(SERVICE_PID_FILTER, servicePid));
+    } catch (Exception e) {
+      throw new SynchronisedInstallerException("Failed to create service filter.", e);
+    }
+  }
+
+  private Dictionary<String, Object> toDic(Map<String, Object> props) {
+    Dictionary<String, Object> dic = new Hashtable<>();
+    props.forEach(dic::put);
+    return dic;
+  }
+
+  private <T> List<T> toList(T ele, T... eles) {
+    return Stream.concat(Stream.of(ele), Stream.of(eles)).collect(Collectors.toList());
+  }
+
+  private <T> Set<T> toSet(T ele, T... eles) {
+    return Stream.concat(Stream.of(ele), Stream.of(eles)).collect(Collectors.toSet());
+  }
+
+  private class ServiceConfigurationListener implements ConfigurationListener {
```

Comment:
✏️ javadoc. I would add a note to make sure this is never a `SynchronizedConfigurationListener` as we want to make sure we are notified from the same pool of threads that are used to dispatch asynchronous configuration events.

Reply:
I am just afraid that the next guy will think it would be better to use a SynchronizedConfigListener. Using this listener over the other was a design choice that is not obvious unless you understand the whole thing and the discussion you and I had on the subject.